src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/UseConcreteTypeAnalyzer.cs

```diff
@@ -0,0 +1,267 @@
+
+        internal static readonly DiagnosticDescriptor UseConcreteTypeForField = DiagnosticDescriptorHelper.Create(
+            RuleId,
+            CreateLocalizableResourceString(nameof(UseConcreteTypeTitle)),
+            CreateLocalizableResourceString(nameof(UseConcreteTypeForFieldMessage)),
+            DiagnosticCategory.Performance,
+            RuleLevel.IdeSuggestion,
+            CreateLocalizableResourceString(nameof(UseConcreteTypeDescription)),
+            isPortedFxCopRule: false,
+            isDataflowRule: false);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(UseConcreteTypeForField);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+
+            if (!Debugger.IsAttached)
+            {
+                context.EnableConcurrentExecution();
+            }
+
+            context.RegisterCompilationStartAction(c =>
+            {
+                var state = new State(c.Compilation);
+                var coll = new Collector(state);
+
+                c.RegisterOperationAction(ctx => coll.HandleInvocation((IInvocationOperation)ctx.Operation), OperationKind.Invocation);
+                c.RegisterOperationAction(ctx => coll.HandleSimpleAssignment((ISimpleAssignmentOperation)ctx.Operation), OperationKind.SimpleAssignment);
+                c.RegisterOperationAction(ctx => coll.HandleDeconstructionAssignment((IDeconstructionAssignmentOperation)ctx.Operation), OperationKind.DeconstructionAssignment);
+                c.RegisterOperationAction(ctx => coll.HandleFieldInitializer((IFieldInitializerOperation)ctx.Operation), OperationKind.FieldInitializer);
+                c.RegisterOperationAction(ctx => coll.HandleVariableDeclarator((IVariableDeclaratorOperation)ctx.Operation), OperationKind.VariableDeclarator);
+                c.RegisterOperationAction(ctx => coll.HandleDeclarationExpression((IDeclarationExpressionOperation)ctx.Operation), OperationKind.DeclarationExpression);
+                c.RegisterOperationAction(ctx => coll.HandleReturn((IReturnOperation)ctx.Operation), OperationKind.Return);
+                c.RegisterCompilationEndAction(ctx => Report(ctx, state));
+            });
+        }
+
+        /// <summary>
+        /// Given all the accumulated analysis state, generate the diagnostics.
+        /// </summary>
+        private static void Report(CompilationAnalysisContext context, State state)
+        {
+            var types = new HashSet<ITypeSymbol>(SymbolEqualityComparer.Default);
+
+            foreach (var field in state.VirtualDispatchFields.Keys)
+            {
+                if (field.DeclaredAccessibility == Accessibility.Public
+                    || field.DeclaredAccessibility == Accessibility.Internal
+                    || field.DeclaredAccessibility == Accessibility.ProtectedOrInternal)
```

Comment:
Alright, this indicates we are only analyzing code assignments within the containing type. We should definitely convert this into a SymbolStart/SymbolEnd analyzer

Reply:
I updated the implementation to reject non-upgradable fields earlier, like we were already doing for methods. This is more consistent.

Reply:
We don't do global analysis, it's all contained in the scope of the single type.